src/serlo-editor-repo/plugin-text/components/link/link-content-search.tsx

```diff
@@ -0,0 +1,273 @@
+
+//TODO: quickbar data only available in de.serlo.org! how should we handle other instances?
+//TODO: quickbar data does not have type for icon
+
+// based on Quickbar, duplicates some code
+
+export function LinkContentSearch({
+  value,
+  setValue,
+  removeLink,
+}: {
+  value: string
+  setValue: (href: string) => void
+  removeLink: () => void
+}) {
+  const [data, setData] = useState<QuickbarData | null>(null)
+  const [query, setQuery] = useState(value)
+  const inputRef = useRef<HTMLInputElement>(null)
+  const [sel, setSel] = useState(-1)
+  const [isEditMode, setIsEditMode] = useState(value.length === 0)
+  const wrapper = useRef<HTMLDivElement>(null)
+  const resultsWrapper = useRef<HTMLDivElement>(null)
+
+  useEffect(() => {
+    setIsEditMode(value.length === 0)
+    setQuery(value)
+  }, [value])
+
+  useEffect(() => {
+    const effectFetch = async () => {
+      const fetchedData = await fetchQuickbarData()
+      if (fetchedData) setData(fetchedData)
+    }
+
+    if (query && !data) {
+      // eslint-disable-next-line no-console
+      effectFetch().catch(console.error)
+    }
+    // eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [query, data])
+
+  useEffect(() => {
+    setSel(0)
+    // setOpen(!!(query && data))
+  }, [query, data, value])
+
+  const results = data ? findResults(data, query) : []
+
+  function setHref(index?: number) {
+    const activeIndex = index ?? sel
+    if (activeIndex < 0 || activeIndex > results.length) return
+
+    const href =
+      activeIndex < results.length ? `/${results[activeIndex].entry.id}` : query
+
+    if (
+      checkSerloHref(href) ||
+      href.startsWith('http://') ||
+      href.startsWith('https://')
+    ) {
+      setValue(href)
+    } else {
+      showToastNotice(
+        'Hmm, das schaut leider nach einem ungültigen Link aus.',
+        'warning'
+      )
+    }
+  }
+
+  const onKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
+    {
+      if (e.key === 'Escape') {
+        setQuery('')
+      }
+      if (e.key === 'ArrowDown' || e.key === 'ArrowUp' || e.key === 'Enter') {
+        if (e.key === 'ArrowDown' && sel < results.length) {
+          setSel(sel + 1)
+        }
+        if (e.key === 'ArrowUp' && sel >= 0) {
+          setSel(sel - 1)
+        }
+        if (e.key === 'Enter') {
+          setHref()
+        }
+        e.preventDefault()
+        return false
+      }
+    }
+  }
+
+  return (
+    <div className="" ref={wrapper}>
+      {isEditMode ? (
+        <>
+          <label className="block px-side pt-4">
+            Suche einen Serlo Inhalt oder füge einen Link ein …
+          </label>
+          <div className="relative w-[27rem]">
+            {renderInput()}
+            {query && renderResetButton()}
+          </div>
+          {renderResults()} <br />
+        </>
+      ) : (
+        <>{renderSetEntry()}</>
+      )}
+    </div>
+  )
+
+  function renderInput() {
```

Comment:
well… this is a code style I adopted while working at serlo and we use this across the whole frontend. 
So changing that would be a massive refactor. 
I now actually like it, since it makes parsing a component a lot easier without spreading code across the repo that is only used in one place.

Reply:
> it's not what we use across the editor codebase

That's a good point. Let's talk about it sometime. 
In general I love to save LOC and really dislike pushing values around the component tree that's why I like this unusual approach but it's also just a habit.

I wanted to split up the big component anyway and did so with **actual** components this time 😸. 
What do you think, code style close enough for now?

Reply:
> I wanted to split up the big component anyway and did so with actual components this time 😸.
What do you think, code style close enough for now?

ping @hejtful and @CodingDive